### ISSUE TITLE
MutationObserver is not defined when running my test

### ISSUE THREAD

**Issue by scote** [thymikee/jest-preset-angular#63]
I use MutationObserver in one component. It seems Jest doesn't know about this ES feature.

I found some documention on Jest website about babel-jest to use preset for ES2015 in `.babelrc` file but I'm not able to make this working with my test.

Here the error I get when running my test.

```
  ReferenceError: MutationObserver is not defined

      at MzSelectDirective.Object.<anonymous>.MzSelectDirective.listenOptionChanges (node_modules/ng2-materialize/dist/select/select.directive.js:172:37)
      at MzSelectDirective.Object.<anonymous>.MzSelectDirective.ngOnInit (node_modules/ng2-materialize/dist/select/select.directive.js:40:14)
      at checkAndUpdateDirectiveInline (node_modules/@angular/core/bundles/core.umd.js:10718:19)
      at checkAndUpdateNodeInline (packages/core/src/view/view.ts:456:1)
      at checkAndUpdateNode (node_modules/@angular/core/bundles/core.umd.js:12065:16)
      at debugCheckAndUpdateNode (packages/core/src/view/services.ts:235:48)
      at debugCheckDirectivesFn (packages/core/src/view/services.ts:294:9)
      at Object.eval [as updateDirectives] (ng:/DynamicTestModule/UserFormComponent.ngfactory.js:1367:5)
      at Object.debugUpdateDirectives [as updateDirectives] (node_modules/@angular/core/bundles/core.umd.js:12620:21)
      at checkAndUpdateView (packages/core/src/view/view.ts:345:12)
      at callViewAction (packages/core/src/view/view.ts:700:1)
      at execComponentViewsAction (node_modules/@angular/core/bundles/core.umd.js:12293:13)
      at checkAndUpdateView (node_modules/@angular/core/bundles/core.umd.js:12038:5)
      at callWithDebugContext (packages/core/src/view/services.ts:645:1)
      at Object.debugCheckAndUpdateView [as checkAndUpdateView] (node_modules/@angular/core/bundles/core.umd.js:12560:12)
      at ViewRef_.Object.<anonymous>.ViewRef_.detectChanges (node_modules/@angular/core/bundles/core.umd.js:10129:63)
      at ComponentFixture.Object.<anonymous>.ComponentFixture._tick (node_modules/@angular/core/bundles/core-testing.umd.js:192:32)
      at packages/core/testing/src/component_fixture.ts:125:1
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone-node.js:365:26)
      at ProxyZoneSpec.Object.<anonymous>.ProxyZoneSpec.onInvoke (node_modules/zone.js/dist/proxy.js:79:39)
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone-node.js:364:32)
      at Object.onInvoke (node_modules/@angular/core/bundles/core.umd.js:4132:37)
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone-node.js:364:32)
      at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/dist/zone-node.js:125:43)
      at NgZone.Object.<anonymous>.NgZone.run (node_modules/@angular/core/bundles/core.umd.js:4001:62)
      at ComponentFixture.Object.<anonymous>.ComponentFixture.detectChanges (packages/core/testing/src/component_fixture.ts:125:1)
      at src/app/users/user-create/user-form/user-form.component.view.spec.ts:65:12
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone-node.js:365:26)
      at ProxyZoneSpec.Object.<anonymous>.ProxyZoneSpec.onInvoke (node_modules/zone.js/dist/proxy.js:79:39)
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone-node.js:364:32)
      at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/dist/zone-node.js:125:43)
      at Object.testBody.length (node_modules/jest-zone-patch/index.js:47:27)
      at new ZoneAwarePromise (node_modules/zone.js/dist/zone-node.js:776:29)
      at Promise.resolve.then.el (node_modules/p-map/index.js:42:16)
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invoke (node_modules/zone.js/dist/zone-node.js:365:26)
      at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/dist/zone-node.js:125:43)
      at node_modules/zone.js/dist/zone-node.js:760:57
      at ZoneDelegate.Object.<anonymous>.ZoneDelegate.invokeTask (node_modules/zone.js/dist/zone-node.js:398:31)
      at Zone.Object.<anonymous>.Zone.runTask (node_modules/zone.js/dist/zone-node.js:165:47)
      at drainMicroTaskQueue (node_modules/zone.js/dist/zone-node.js:593:35)
      at ZoneTask.invoke (node_modules/zone.js/dist/zone-node.js:464:25)
      at timer (node_modules/zone.js/dist/zone-node.js:1638:29)
      at Timeout.callback [as _onTimeout] (node_modules/jsdom/lib/jsdom/browser/Window.js:523:19)
      at ontimeout (timers.js:386:14)
      at tryOnTimeout (timers.js:250:5)
      at Timer.listOnTimeout (timers.js:214:5)
```

**Comment by thymikee** [comment]
It's not about Jest, but JSDom. You need to provide a polyfill, like the one for localStorage here: https://github.com/thymikee/jest-preset-angular/blob/master/example/src/jestGlobalMocks.ts

**Comment by scote** [comment]
I was able to make Jest working with MutationObserver. Here my solution if someone need to do the same thing.

I added this polyfill to my `mocks` folder :
https://github.com/aurelia/pal-nodejs/blob/master/src/polyfills/mutation-observer.ts

And added a new define property for window.

jestGlobalMocks.ts
```
import { MutationObserver } from './polyfills/mutation-observer';

const storageMock = () => {
  let storage = {};
  return {
    getItem: key => key in storage ? storage[key] : null,
    setItem: (key, value) => storage[key] = value || '',
    removeItem: key => delete storage[key],
    clear: () => storage = {},
  };
};

Object.defineProperty(window, 'localStorage', { value: storageMock() });
Object.defineProperty(window, 'sessionStorage', { value: storageMock() });
Object.defineProperty(window, 'MutationObserver', { value: MutationObserver });
Object.defineProperty(window, 'getComputedStyle', {
  value: () => ['-webkit-appearance'],
});
```

**Comment by thymikee** [comment]
Thanks for followup! 🙂

**Comment by dudewad** [comment]
I know I'm a little late to this, but I can't for the life of me figure out how to get polyfills to actually work. Nothing in the Jest documentation makes it clear to me. I see that `setupFilesAfterEnv` is now used in place of `setupTestFrameworkScriptFile` but neither have any effect for me. Does anyone have a link to some documentation where I can learn about setting up polyfills?

**Comment by thymikee** [comment]
maybe you can try latest alpha version (7.0.0-something)? We have a newest jsdom there, and they recently added mutation observer (not sure if released though, see https://github.com/jsdom/jsdom/pull/239)

**Comment by ahnpnl** [comment]
I haven't tested loading polyfills into jest environment yet. I saw by default Angular CLI adds polyfills.ts to tsconfig.spec.json. Not sure if that file is loaded properly. That is also a way to provide extra polyfills if needed.

And I think setupFilesAfterEnv can also be used to import polyfills right @thymekee ?

**Comment by thymikee** [comment]
Yea, this or `setupEnv` (it loads earlier)

**Comment by dudewad** [comment]
@thymikee @ahnpnl 

Some detail: We're running an nrwl nx workspace. We have multiple libraries that all run out of the same repository. Our Jest configs all use the one in the project root as a preset. I tried setting `setupEnv` and `setupFilesAfterEnv`. `setupEnv` complains that it can't find the polyfills file because even with `<rootDir>` it looks from the root of the lib, not from the root of the preset file (despite these props being set in the rood preset file).

`setupFilesAfterEnv` appears to have no effect whatsoever.

I got it to work by specifying the global override in the spec file itself (I had to use `global['MutationObserver']` instead of `global.MutationObserver` because my IDE complains that `global` doesn't contain that property). But this solution is undesirable because the component I have that utilizes `MutationObserver` is low-level and used in lots of libraries. Patching it on a per-suite basis is "un-wonderful".

I can provide more details if you want. I imagine that I am just messing something up with configuration...

FWIW I appreciate your help in suggesting using the latest alpha but I would rather resolve the issue of polyfills not working for the likely future case where I'll want to add another one. I will look into the tsconfig.spec.json as @ahnpnl suggested, and see if I can figure out what I've done wrong.

Thanks!
